### ISSUE TITLE
Ticket #2240 Background jobs queue (aka: file copy queue)

### ISSUE THREAD

**Issue by Gilwyad** [MidnightCommander/mc-old#8]
Hi, there is a two year old open ticket with a patch here: https://www.midnight-commander.org/ticket/2240

It makes sure that there is only one active background copy operation, by automatically stopping and starting these background jobs.

For your information, I applied the patch to version 4.7.4 and verified that it's working, then merged into master without conflicts. It was again working flawlessly. 

I'd like to see this patch applied to Midnight Commander.

**Comment by slavaz** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA1

03.07.2012 21:36, Peter wrote:

Hi Peter,

> Hi, there is a two year old open ticket with a patch here:
> https://www.midnight-commander.org/ticket/2240 It makes sure that
> there is only one active background copy operation, by
> automatically stopping and starting these background jobs.

Yes, the queue for background tasks sounds good for me, but other
developers disagree with the only one active background process
because background operations should be completed fast as possible,
they said. On boxes with milticore/miltiple CPU's or in case when
files copied from/to different storages, these arguments sounds
convincing.

So, from my side I want to propose a configure option for toggle
queue/parallel algorithm in background operations. I hope that this
solution will bring together the fans of background queue and the
parallel background operations.

---

WBR, Slavaz.

-----BEGIN PGP SIGNATURE-----
Version: GnuPG v1.4.12 (GNU/Linux)
Comment: Using GnuPG with Mozilla - http://enigmail.mozdev.org/

iEYEARECAAYFAk/zd2EACgkQb3oGR6aVLpodbQCfYQXwU/G+3PoCzkzy8xucO7wG
3DIAn0Eczfdhg0aOdOcFPrTI46E1nKfv
=HtCL
-----END PGP SIGNATURE-----

**Comment by Gilwyad** [comment]
On Wed, Jul 4, 2012 at 12:51 AM, Slava Zanko
reply@reply.github.com
wrote:

> So, from my side I want to propose a configure option for toggle
> queue/parallel algorithm in background operations. I hope that this
> solution will bring together the fans of background queue and the
> parallel background operations.

Yes, a configuration option would be great! Can I help somehow to make
this happen soon? I'm new to git, github and C :)

**Comment by slavaz** [comment]
Em... sorry, you don't :)

This task should be completed by one of developers.

**Comment by zatricky** [comment]
@slavaz, the whole point of open source is that new people come in, learn, and contribute. Try to be more encouraging.

That said, @Gilwyad, a mature project is usually the wrong type of project to take on as a "junior". Maybe things have changed for you since. I'd suggest following the development of an active project to get an idea of how things work, and to learn from that.

@slavaz mentioned (a long long time ago):
"other developers disagree with the only one active background process because background operations should be completed fast as possible"
The above makes sense with SSDs or with memory for _performance_. With spinning rust, parallel operations result in _MUCH WORSE_ performance. With SSDs, it is also arguable that parallel operations result in _slightly_ more wear (overall, probably less than 5% extra).

On the spinning-rust parallel performance issue, the simple fact is that any SATA drive (spindle or SSD) only performs operations sequentially. Additionally they are limited in the number of operations they perform per second. Spindles are typically up to 150 IOps (I/O Operations Per Second) while SSDs are typically between 40k and 1M IOps (the real reason they're subjectively faster than spindles).

From the above, if you attempt to read as many far-apart 1KB files from a spindle in one second, you will typically read between 70 and 200 (depending on if you have a low-end consumer drive or a high-end enterprise drive). If you expand that to be that you want to read 150 files simultaneously, you are likely to read only a single chunk of each file - and it will take you an _entire_ second to read one chunk. One chunk is typically 4KB, thus you are now reading at an abysmal 600KB/s.

That 600KB/s is SUPER slow. If you tried to copy 1GB files, 150 of them, simultaneously, it will take you two days, 21 hours, and 27 minutes. If you instead copy them sequentially (one at a time), you will be doing 1 IOps (sequentially) but at the maximum speed of the spindle. Let's say it can do 80MB/s. You will average slightly less than that, but still very close to maximum, and it will probably take ~ 31 minutes.

You can easily test this theory within Linux or Windows:
Set up 10 operations to run in parallel and sequentially. These copy operations must be from a spindle drive and must be copied to /dev/null, a ramdisk, or an SSD. The files must preferably be large enough that each takes more than a second to copy (200MB or larger is a good starting point).

Wipe the caches (or reboot), and execute the copies in sequence. You will find the average copy speed of each one is relatively stable and close to the drive's maximum speed.

Wipe the caches again (or reboot), and execute the copies in parallel. You will find that the average copy speed of each one is _less than_ a tenth of the copy speed seen previously. Thus the total speed will be less, and the copy operation will take longer to complete.

**Comment by igitur** [comment]
If you need a more concrete use case, I compared 2 scenarios: 

1) Copy 100GB from internal HDD to external HDD (USB3) by queueing the data.
2) Copy the same data from same source to same destination, but batch it in 5 20GB chunks.

The first option is definitely faster. I suspect the overhead is exaggerated on an external hard drive. My gut feel is that for internal copying (internal HDD to another internal HDD) the r/w overhead of simultaneous background tasks wouldn't be that much.

**Comment by zyv** [comment]
https://www.midnight-commander.org/ticket/2240#comment:15

**Comment by igitur** [comment]
I see this is closed, but does that mean the ticket is resolved or won't be?

**Comment by zyv** [comment]
This only means that issues are to be tracked in Trac, and nobody follows this tracker. Unfortunately, it was enabled when then repository was created, and since then, some issues have been created here. I'm moving all relevant content to the Trac and will disable this tracker, so that nobody wastes their time to report issues here only to be ignored for years...
